crates/sui-json-rpc/src/error.rs

```diff
@@ -76,19 +80,89 @@ impl Error {
-                    let error_object =
-                        ErrorObject::owned(-32000, NON_RECOVERABLE_ERROR_MSG, Some(errors));
-                    RpcError::Call(CallError::Custom(error_object))
-                }
-                _ => RpcError::Call(CallError::Failed(err.into())),
-            },
+            Error::QuorumDriverError(err) => match_quorum_driver_error(err),
             _ => RpcError::Call(CallError::Failed(self.into())),
         }
     }
 }
 
+pub fn match_quorum_driver_error(err: QuorumDriverError) -> RpcError {
+    match err {
+        QuorumDriverError::InvalidUserSignature(err) => {
+            let inner_error_str = match err {
+                // Use inner UserInputError's Display instead of SuiError::UserInputError, which renders UserInputError in debug format
+                SuiError::UserInputError { error } => error.to_string(),
+                _ => err.to_string(),
+            };
+
+            let error_message = format!("Invalid user signature: {inner_error_str}");
+
+            let error_object = ErrorObject::owned(INVALID_PARAMS_CODE, error_message, None::<()>);
+            RpcError::Call(CallError::Custom(error_object))
+        }
+        QuorumDriverError::TimeoutBeforeFinality
+        | QuorumDriverError::FailedWithTransientErrorAfterMaximumAttempts { .. } => {
+            RpcError::Call(CallError::InvalidParams(err.into()))
+        }
+        QuorumDriverError::ObjectsDoubleUsed {
+            conflicting_txes,
+            retried_tx,
+            retried_tx_success,
+        } => {
+            let error_message = format!(
+                "Failed to sign transaction by a quorum of validators because of locked objects. Please review the 'data' field for the conflicting transactions and associated objects. Retried a conflicting transaction {:?}, success: {:?}",
+                retried_tx,
+                retried_tx_success
+            );
+
+            let mut new_map: BTreeMap<TransactionDigest, Vec<ObjectRef>> = BTreeMap::new();
+
+            for (digest, (pairs, _)) in conflicting_txes {
+                let mut new_vec = Vec::new();
+
+                for (_authority, obj_ref) in pairs {
+                    new_vec.push(obj_ref);
+                }
+
+                new_map.insert(digest, new_vec);
+            }
+
+            let error_object =
+                ErrorObject::owned(INVALID_PARAMS_CODE, error_message, Some(new_map));
+            RpcError::Call(CallError::Custom(error_object))
+        }
+        QuorumDriverError::NonRecoverableTransactionError { errors } => {
+            let error_message = format!("{NON_RECOVERABLE_ERROR_MSG} - please review the 'data' field for the errors that occurred during execution, correct the issues if applicable, and try again.");
+            let new_errors: Vec<String> = errors
+                .into_iter()
+                .map(|(err, _, _)| {
+                    match err {
```

Comment:
as discussed elsewhere, i think it's helpful to exclude transient errors such as SuiError::RpcError. Those are not actionable to clients and could it more confusing